src/main/java/io/spokestack/spokestack/tts/SpokestackTTSClient.java

```diff
@@ -54,102 +70,138 @@ public SpokestackTTSClient(TTSCallback callback) {
+                param = "ssml";
+                method = "synthesizeSsml";
+                break;
+            default:
+                break;
+        }
+        variables.put("voice", request.voice);
+        String queryString = String.format(GRAPHQL_QUERY, param, method);
+        postSpeech(headers, queryString, variables);
     }
 
-    private void postSpeech(HashMap<String, String> speechBody) {
-        if (this.ttsApiKey == null) {
-            ttsCallback.onError("API key not provided");
+    private void postSpeech(Map<String, String> headers,
+                            String queryString,
+                            Map<String, String> variables) {
+        if (this.ttsClientKey == null) {
+            ttsCallback.onError("client key not provided");
+            return;
+        }
+        if (this.ttsClientSecret == null) {
+            ttsCallback.onError("client secret not provided");
             return;
         }
         if (this.ttsUrl == null) {
             ttsCallback.onError("TTS URL not provided");
             return;
         }
 
-        RequestBody postBody = RequestBody.create(APPLICATION_JSON,
-              gson.toJson(speechBody));
+        String bodyJson = gson.toJson(variables);
+        String fullBody =
+              "{\"query\": \"" + queryString + "\", "
+                    + "\"variables\": " + bodyJson + "}";
+        RequestBody postBody = RequestBody.create(fullBody, APPLICATION_JSON);
 
-        Request request = new Request.Builder()
+        Request.Builder builder = new Request.Builder();
+
+        for (Map.Entry<String, String> header : headers.entrySet()) {
+            builder = builder.addHeader(header.getKey(), header.getValue());
+        }
+
+        String authHeader = signRequest(fullBody);
+        if (authHeader == null) {
+            // the error is dispatched to the callback by signRequest
+            return;
+        }
+
+        Request request = builder
               .url(ttsUrl)
-              .header("content-type", "application/json")
-              .header("Authorization", "Key " + this.ttsApiKey)
+              .header("Content-Type", "application/json")
+              .header("Authorization",
+                    "Spokestack " + ttsClientKey + ":" + authHeader)
               .post(postBody)
               .build();
 
         httpClient.newCall(request).enqueue(this.ttsCallback);
     }
 
+    String signRequest(String body) {
+        String base64Signature = null;
+        try {
+            Mac hmacAlgo = Mac.getInstance(HMAC_TYPE);
+            byte[] keyBytes =
+                  this.ttsClientSecret.getBytes(StandardCharsets.UTF_8);
+            SecretKeySpec keySpec = new SecretKeySpec(keyBytes, HMAC_TYPE);
+            hmacAlgo.init(keySpec);
+            byte[] macData = hmacAlgo.doFinal(
+                  body.getBytes(StandardCharsets.UTF_8));
+            base64Signature = Base64.encode(macData);
+        } catch (NoSuchAlgorithmException e) {
```

Comment:
might be my favorite error you had to deal with

Reply:
#agree